File: INTRODUCCION-POO/EJERCICIO-7/Celular.py
# INTRODUCCIÓN A LA PROGRAMACIÓN ORIENTADA A OBJETOS EN PYTHON

# Definición de la clase Celular
class Celular:
    # Constructor de la clase
    def __init__(self):
        self.espacio_total = 1024  # Espacio total en MB
        self.espacio_ocupado = 0   # Espacio ocupado en MB
        self.bateria = 100         # Batería inicial en 100%
        self.aplicaciones = []     # Lista de aplicaciones instaladas

    # Método para instalar una nueva aplicación
    def instalar_aplicacion(self, nombre, tamaño):
        if self.espacio_ocupado + tamaño <= self.espacio_total:
            self.aplicaciones.append({"nombre": nombre, "tamaño": tamaño})
            self.espacio_ocupado += tamaño
            print(f"Aplicación '{nombre}' instalada. Espacio ocupado: {self.espacio_ocupado} MB")
        else:
            print(f"No hay suficiente espacio para instalar '{nombre}'.")

    # Método para utilizar una aplicación
    def usar_aplicacion(self, nombre, minutos):
        if self.bateria <= 0:
            print("Celular apagado. No se puede usar ninguna aplicación.")
            return

        for app in self.aplicaciones:
            if app["nombre"] == nombre:
                tamaño = app["tamaño"]
                if tamaño > 250:
                    consumo = 5 * (minutos / 10)
                elif tamaño > 100:
                    consumo = 2 * (minutos / 10)
                else:
                    consumo = 1 * (minutos / 10)

                self.bateria -= consumo
                if self.bateria < 0:
                    self.bateria = 0
                    print(f"Celular apagado. Batería agotada al usar '{nombre}'.")
                else:
                    print(f"Usando '{nombre}' por {minutos} minutos. Batería restante: {self.bateria:.2f}%")
                return
        print(f"Aplicación '{nombre}' no encontrada.")

    # Método para mostrar el porcentaje de batería restante
    def mostrar_bateria(self):
        print(f"Batería restante: {self.bateria:.2f}%")


# Crear un objeto de la clase Celular
mi_celular = Celular()

# Instalar aplicaciones
mi_celular.instalar_aplicacion("WhatsApp", 50)
mi_celular.instalar_aplicacion("Instagram", 120)
mi_celular.instalar_aplicacion("Juego Pesado", 300)

# Usar aplicaciones
mi_celular.usar_aplicacion("WhatsApp", 20)
mi_celular.usar_aplicacion("Instagram", 30)
mi_celular.usar_aplicacion("Juego Pesado", 10)

# Mostrar batería restante
mi_celular.mostrar_bateria()

File: INTRODUCCION-POO/EJERCICIO-7/Celular.java
// INTRODUCCIÓN A LA PROGRAMACIÓN ORIENTADA A OBJETOS EN JAVA

import java.util.ArrayList;
import java.util.List;


 public class Celular {
    
    private int espacioTotal = 1024;  
    private int espacioOcupado = 0;   
    private double bateria = 100;     
    private List<Aplicacion> aplicaciones = new ArrayList<>();  

    
    private class Aplicacion {
        String nombre;
        int tamaño;

        public Aplicacion(String nombre, int tamaño) {
            this.nombre = nombre;
            this.tamaño = tamaño;
        }
    }


    public void instalarAplicacion(String nombre, int tamaño) {
        if (espacioOcupado + tamaño <= espacioTotal) {
            aplicaciones.add(new Aplicacion(nombre, tamaño));
            espacioOcupado += tamaño;
            System.out.println("Aplicación '" + nombre + "' instalada. Espacio ocupado: " + espacioOcupado + " MB");
        } else {
            System.out.println("No hay suficiente espacio para instalar '" + nombre + "'.");
        }
    }

    
    public void usarAplicacion(String nombre, int minutos) {
        if (bateria <= 0) {
            System.out.println("Celular apagado. No se puede usar ninguna aplicación.");
            return;
        }

        for (Aplicacion app : aplicaciones) {
            if (app.nombre.equals(nombre)) {
                double consumo;
                if (app.tamaño > 250) {
                    consumo = 5 * (minutos / 10.0);
                } else if (app.tamaño > 100) {
                    consumo = 2 * (minutos / 10.0);
                } else {
                    consumo = 1 * (minutos / 10.0);
                }

                bateria -= consumo;
                if (bateria < 0) {
                    bateria = 0;
                    System.out.println("Celular apagado. Batería agotada al usar '" + nombre + "'.");
                } else {
                    System.out.printf("Usando '%s' por %d minutos. Batería restante: %.2f%%\n", nombre, minutos, bateria);
                }
                return;
            }
        }
        System.out.println("Aplicación '" + nombre + "' no encontrada.");
    }

    // Método para mostrar el porcentaje de batería restante
    public void mostrarBateria() {
        System.out.printf("Batería restante: %.2f%%\n", bateria);
    }


    public static void main(String[] args) {
        Celular miCelular = new Celular();

        // Instalar aplicaciones
        miCelular.instalarAplicacion("WhatsApp", 50);
        miCelular.instalarAplicacion("Instagram", 120);
        miCelular.instalarAplicacion("Juego Pesado", 300);

        // Usar aplicaciones
        miCelular.usarAplicacion("WhatsApp", 20);
        miCelular.usarAplicacion("Instagram", 30);
        miCelular.usarAplicacion("Juego Pesado", 10);

        // Mostrar batería restante
        miCelular.mostrarBateria();
    }
}